### ISSUE TITLE
possible ordering error among a random_pet / template_file / (locals / aws_instance / null_resource)

### ISSUE THREAD

**Issue by novas0x2a** [hashicorp/terraform#16441]
### Terraform Version
Terraform v0.10.7

### Terraform Configuration Files
```hcl
resource "random_pet" "foo" {}
data "template_file" "foo" {
  template = "wee says hi!"
  vars {
    hostname = "foo-${random_pet.foo.id}"
  }
}
locals {
  hostnames = "${data.template_file.foo.vars.hostname}"
}
output "hostnames" {
  value = "${data.template_file.foo.vars.hostname}"
}
```

### Debug Output
https://gist.github.com/novas0x2a/a5342fe5cf57b68075461d663fd2db46

### Expected Behavior
Applied successfully

### Actual Behavior
```
Error running plan: 1 error(s) occurred:

* local.hostnames: local.hostnames: Resource 'data.template_file.foo' does not have attribute 'vars.hostname' for variable 'data.template_file.foo.vars.hostname'
```

### Steps to Reproduce
Please list the full steps required to reproduce the issue, for example:
1. `terraform init`
2. `terraform apply`

### Important Factoids
This is a reduction of a larger module in which the interpolation in the locals block is actually in the tags block of an aws_instance, so I don't think this is necessarily a locals-specific problem (if i use a null_resource, I see the same error).

If I comment out the locals block, run apply, then uncomment the locals block and run apply again, it will succeed. I'm assuming this means that the access to the vars block is being scheduled before the final calculation of the vars block.

### References
- This sounds pretty similar to #15568 except null_resource doesn't solve the problem and I'm using 0.10.7 where this fix has already landed.

**Comment by apparentlymart** [comment]
Thanks for reporting this @novas0x2a, and sorry for the strange behavior.

Some notes from the debug log for some deeper analysis later....

This is the graph just before the transitive reduction transform, showing that the expected dependency edges between the three nodes are there:

```
data.template_file.foo - *terraform.NodePlannableResource
  provider.template - *terraform.NodeApplyableProvider
  random_pet.foo - *terraform.NodePlannableResource
local.hostnames - *terraform.NodeLocal
  data.template_file.foo - *terraform.NodePlannableResource
meta.count-boundary (count boundary fixup) - *terraform.NodeCountBoundary
  data.template_file.foo - *terraform.NodePlannableResource
  local.hostnames - *terraform.NodeLocal
  output.hostnames - *terraform.NodeApplyableOutput
  provider.random - *terraform.NodeApplyableProvider
  provider.template - *terraform.NodeApplyableProvider
  random_pet.foo - *terraform.NodePlannableResource
output.hostnames - *terraform.NodeApplyableOutput
  data.template_file.foo - *terraform.NodePlannableResource
provider.random - *terraform.NodeApplyableProvider
provider.random (close) - *terraform.graphNodeCloseProvider
  provider.random - *terraform.NodeApplyableProvider
  random_pet.foo - *terraform.NodePlannableResource
provider.template - *terraform.NodeApplyableProvider
provider.template (close) - *terraform.graphNodeCloseProvider
  data.template_file.foo - *terraform.NodePlannableResource
  provider.template - *terraform.NodeApplyableProvider
random_pet.foo - *terraform.NodePlannableResource
  provider.random - *terraform.NodeApplyableProvider
root - terraform.graphNodeRoot
  meta.count-boundary (count boundary fixup) - *terraform.NodeCountBoundary
  provider.random (close) - *terraform.graphNodeCloseProvider
  provider.template (close) - *terraform.graphNodeCloseProvider
```

Skimming the log for the "entering eval tree" and "exiting eval tree" nodes helps us see what order things were evaluated in, and which things were evaluated in parallel:

```
Entering eval tree: provider.random
Entering eval tree: provider.template
Exiting eval tree: provider.random
Entering eval tree: random_pet.foo
Exiting eval tree: random_pet.foo
Exiting eval tree: provider.template
Entering eval tree: random_pet.foo
Exiting eval tree: random_pet.foo
Entering eval tree: data.template_file.foo
Entering eval tree: provider.random (close)
Exiting eval tree: provider.random (close)
Exiting eval tree: data.template_file.foo
(data.template_file.foo dynamic-expands to produce data.template_file.foo[0])
Entering eval tree: data.template_file.foo[0]
Exiting eval tree: data.template_file.foo[0]
Entering eval tree: local.hostnames
Entering eval tree: provider.template (close)
Entering eval tree: output.hostnames
Exiting eval tree: output.hostnames
Exiting eval tree: local.hostnames
Exiting eval tree: provider.template (close)
root: eval: *terraform.EvalLocal, err: local.hostnames: Resource 'data.template_file.foo' does not have attribute 'vars.hostname' for variable 'data.template_file.foo.vars.hostname'
Entering eval tree: output.hostnames
Exiting eval tree: output.hostnames
Exiting eval tree: local.hostnames
Exiting eval tree: provider.template (close)
```

This suggests that the walk is happening in the correct order, so something more tricky seems to be going on here. My unsubstantiated theory is that something is not working quite right in the code that looks up values from the state, possibly a misinterpretation of the map that's recorded there, causing it to seem like the map as a whole or the specific key doesn't exist.

We're doing some other work in this area right now for the next-generation config parser so it'll be interesting to see if this reproduces in the new codepath that does the translation from state to interpolation in a slightly different way. I'll pause looking at this for the moment and return to it once I can try both codepaths and look for differences.

Thanks for the great reproduction case, @novas0x2a!

**Comment by sublee** [comment]
I have a similar case.  I'll subscribe this issue.

**Comment by joe-boyce** [comment]
Is this also related to the following issue:

```
module.xxxx.output.tag: Resource 'random_pet.tag' does not have attribute 'keepers.tag' for variable 'random_pet.tag.keepers.tag'
```

**Comment by joe-boyce** [comment]
Any update on this one?

**Comment by SamWhited** [comment]
I am seeing something very similar to this on v0.11.7 except with a `random_id` where it thinks that nothing exists in keepers. If anyone has come up with a workaround, I'd lvoe to hear it.

```
Terraform v0.11.7
+ provider.cloudflare v1.0.0
+ provider.fastly v0.2.0
+ provider.random v1.3.1
+ provider.tls v1.1.0
+ provider.triton v0.5.1
```

**Comment by joe-boyce** [comment]
Any updates on this one?

**Comment by jbardin** [comment]
I've confirmed this is working as expected in 0.12.
If there are similar issues with 0.12, I suspect they will from a different root cause, and should be opened in a new issue.

Thanks!
